### PR TITLE
Continue to test autoresearch

### PR DESCRIPTION
**Pull request by mrvictoru**
Continue to test and refine using autoresearch

### CODE REVIEW

**Comment by mrvictoru** [comment]
### Would a code edit auto research loop be more effective?
**Potentially yes**, because editing training code expands the search space beyond hyperparameters:
- loss formulation
- batching/sampling
- normalization
- scheduler logic
- target construction
- architecture wiring
- prompt construction
- masking behavior
- eval-time inference logic tied to training assumptions

That can unlock improvements config search simply cannot reach.

### Would it be possible here?
**Yes, technically very possible.**  
This repo already has the right building blocks:
- CLI training entrypoints
- isolated eval scripts
- autoresearch runner / ledger / stage gates
- tests
- frozen benchmarks

So the repo is actually in a decent position to support a “training-code-edit” harness.

### Should you do it now?
**Probably not as the first next step.**  
I’d recommend:
1. prove the current config-only loop works on real runs
2. then add a **restricted training-code mutation mode**
3. keep config-only as the safe baseline

---

# Why Karpathy-style code-edit search can be stronger

Config-only search assumes:
- current code structure is already basically right
- best gains come from knob tuning

That’s often false.

Code-edit search can explore things like:
- changing RTG conditioning
- altering return scaling logic
- reweighting action/state/return losses dynamically
- changing sequence packing or context truncation
- curriculum or subset sampling
- optimizer and warmup schedule behavior
- gradient clipping defaults
- mixed precision safeguards
- better checkpoint selection criteria
- validation metric selection
- train/val split policy

In a DT project, those can matter **more** than moving `lr` from `2e-5` to `5e-5`.

So in principle, **yes, a code-edit harness is a more expressive and often more powerful search regime**.

---

# Why it is riskier in this repo

The current autoresearch design is deliberately constrained:
- frozen benchmark
- allowed mutable keys
- keep/discard logic
- no notebook mutation
- no dataset mutation
- no reward mutation

That makes it:
- auditable
- reproducible
- relatively safe

If you allow direct training-code edits, the agent can accidentally:
- break compatibility with saved checkpoints
- silently alter evaluation assumptions
- introduce data leakage
- mutate code paths outside “training only”
- optimize for benchmark quirks rather than generalization
- produce improvements that are hard to interpret

The main danger is **search-space explosion**.

With config-only:
- each trial is easy to explain

With code-edit:
- each trial may change multiple coupled behaviors
- blame/credit becomes harder
- reproducibility becomes much more important

So it can be more effective, but only if tightly sandboxed.

---

# Is it possible in this codebase specifically?

## Yes, because the repo already has most prerequisites

### 1. Training entrypoints are already isolated
You have:
- `src/pretrain_decision_transformer.py`
- `src/pretrain_aemo_decision_transformer.py`

Those are natural boundaries for “editable training code”.

### 2. Evaluation is already separated
You have:
- `src/eval_household.py`
- `src/eval_aemo.py`
- `src/eval_common.py`

That’s exactly what you want.  
The harness can mutate training code while keeping evaluation frozen.

### 3. Runner/ledger/stage gates already exist
You already have:
- `runner.py`
- `stage_a.py`
- `stage_b.py`
- `ledger.py`

These can be extended to log:
- patch diffs
- edited files
- code hash
- parent run
- failure reason

### 4. Tests exist
There’s already autoresearch-related test coverage, which helps if you later add:
- “edited file must still pass import smoke”
- “edited code must preserve CLI contract”
- “edited code cannot touch eval modules”

So yes, this repo is structurally compatible with a Karpathy-style harness.

---

# What “possible” would mean here

A practical version in this repo would look like:

## Editable files allowed
Only allow edits under a narrow whitelist such as:
- `src/pretrain_decision_transformer.py`
- `src/pretrain_aemo_decision_transformer.py`
- maybe `src/transformer_training.py`

Possibly also:
- `src/decision_transformer.py`

But I would start **without** allowing model architecture file edits unless you really want that.

## Frozen files
Never allow edits to:
- `src/eval_*`
- `src/autoresearch/*` except maybe patch application internals
- benchmark config JSONs
- env code
- reward/degradation logic
- notebooks
- tests unless specifically in a separate mode

That preserves fair comparison.

---

# Would it actually be more effective here?

## Likely yes for these areas

In this repo, the most promising code-level search targets are probably:

### 1. Loss shaping
The training code likely exposes a lot of room for:
- balancing action/state/return losses
- weighting timesteps differently
- masking padded tokens differently
- emphasizing late trajectory segments
- auxiliary penalties

### 2. Data sampling
Potential improvements:
- better episode sampling
- priority on high-return / in-distribution windows
- length-aware batching
- curriculum over context lengths
- subset strategy for large AEMO corpora

### 3. RTG / prompt handling
This repo already mentions:
- `recommended_rtg`
- `recommended_return_scale`
- prompt calibration

That’s a sign that **training/inference prompt mechanics are important**, and those often benefit from code changes more than config changes.

### 4. Stability logic
Things like:
- AMP guardrails
- gradient clipping
- NaN handling
- checkpoint selection
- early stopping
- LR warmup/decay
- resume semantics

These are often code-level improvements.

So yes, especially for DT training, a training-code-edit loop could plausibly outperform a pure hyperparameter loop.

---

# The main question: should the harness edit code directly or generate patches?

It should **generate patches**, not arbitrary file rewrites.

Best design:
- harness proposes unified diff or structured patch
- patch applied to a temporary worktree
- run import smoke tests
- run Stage A / Stage B
- record patch in ledger
- keep/discard based on evaluation

That is much safer than “rewrite whole file”.

---

# Recommended design for this repo

## Option A: Restricted patch search over training files only
This is the most realistic next evolution.

### Allowed targets
Start with:
- `src/transformer_training.py`
- `src/pretrain_decision_transformer.py`
- `src/pretrain_aemo_decision_transformer.py`

### Optional later
- `src/decision_transformer.py`

### Hard restrictions
- max 1–2 files changed per trial
- max patch size
- only training-related symbols
- no new dependencies
- no network access
- no changes to evaluation files
- no changes to benchmarks
- no changes to tests in search mode

This is very implementable.

---

## Option B: “Template slots” instead of open-ended code editing
This may actually be the best compromise.

Instead of letting the agent edit arbitrary code, define explicit extension points like:
- loss function mode
- sampler mode
- RTG calibration mode
- scheduler mode
- checkpoint selection mode

Then the agent edits only those plugin modules or switches between registered strategies.

This gives you:
- much more expressivity than config-only
- much less risk than free-form code editing

In many real systems, this is the sweet spot.

---

# What would need to be added to implement this here

## 1. Trial workspace isolation
Right now runner seems oriented around configs/artifacts.  
For code-edit search, you’d want:
- temp copy / git worktree per run
- apply patch in isolated directory
- run train/eval there
- delete or keep artifact snapshot

This is very important.

## 2. Patch logging in ledger
Add fields like:
- `edited_files`
- `patch_diff`
- `base_commit_sha`
- `trial_code_sha256`
- `parent_run_id`

Without this, reproducibility gets weak fast.

## 3. Static safety checks
Before training:
- patch touches only allowed files
- patch does not import forbidden modules
- patch does not modify eval harness
- patch size under threshold
- files still parse/import

## 4. Contract tests
Run quick checks before Stage A:
- training CLI still accepts expected args
- model save/checkpoint outputs still exist
- eval can still load produced model

## 5. Patch proposer backend
Instead of JSON config output, the LLM would output either:
- a unified diff
- or structured edit instructions

I’d prefer **structured edit instructions** first, because parsing and safety checks are easier.

Example:
```json
{
  "file": "src/transformer_training.py",
  "edits": [
    {
      "symbol": "train_decision_transformer",
      "operation": "insert_after",
      "anchor": "loss.backward()",
      "code": "torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)"
    }
  ],
  "rationale": "Add gradient clipping to improve stability."
}
```

That’s safer than raw free-form patch text.

---

# A realistic phased plan

## Phase 1
Keep current config-only system and actually run it.

## Phase 2
Add **restricted training-code patch mode** with:
- isolated worktree
- editable file whitelist
- patch recorded in ledger
- import smoke tests

## Phase 3
Add **mixed search**
- first propose config mutation
- then optionally code patch
- compare against current best

## Phase 4
Refactor common “interesting mutation surfaces” into explicit strategy plugins

This would likely be the most maintainable long-term architecture.

---

# My opinion on effectiveness

If your goal is:

### “Find quick, safe, reproducible improvements”
Then **config-only** is better first.

### “Find the best possible improvements in DT training”
Then **restricted training-code search** will probably be more effective.

In this repo, I suspect the biggest gains may come from:
- RTG calibration logic
- loss shaping
- data sampling strategy
- training stability measures

Those are all more code-level than hyperparameter-level.

So my honest answer is:

**Yes, a Karpathy-style training-code-edit harness is probably a stronger optimization method for this repo than config-only search, provided you restrict the editable surface carefully.**

---

# My recommendation for this codebase

I would not replace the current system. I would add a second mode:

- `--search-space config`
- `--search-space training-code`

Where `training-code` means:
- only whitelisted files
- only isolated temporary worktree
- frozen eval
- patch + code hash recorded in ledger
- pre-Stage-A import/contract checks

That would be both feasible and useful.

---

# Best editable surface to start with here

If you want the safest first version, I’d allow edits only in:

1. `src/transformer_training.py`
2. `src/pretrain_decision_transformer.py`
3. `src/pretrain_aemo_decision_transformer.py`

And explicitly disallow edits to:
- envs
- eval scripts
- helper metrics
- benchmarks
- autoresearch harness itself

That gives you a clean “training code only” search space similar in spirit to the Karpathy setup.

---

# Bottom line

## Would it be more effective?
Yes, likely more effective than config-only search for DT tuning.

## Would it be possible here?
Yes, definitely possible in this repo.

## Would I recommend it?
Yes — but as a **restricted, second-mode implementation**, not as an uncontrolled free-form replacement.

If you want, I can turn this into a concrete design proposal for this repo:  
**files to add/change, ledger schema changes, safety rules, and a minimal `training-code` mode architecture.**
